spark3/src/main/java/org/apache/iceberg/actions/Spark3SnapshotAction.java

```diff
@@ -0,0 +1,141 @@
+                       Identifier destTableName) {
+    super(spark, sourceCatalog, sourceTableName, destCatalog, destTableName);
+    // Cannot check if the table location that would be generated by the catalog will match the source
+    this.destTableLocation = null;
+  }
+
+  Spark3SnapshotAction(SparkSession spark, CatalogPlugin sourceCatalog,
+                       Identifier sourceTableName, CatalogPlugin destCatalog,
+                       Identifier destTableName, String destTableLocation) {
+    super(spark, sourceCatalog, sourceTableName, destCatalog, destTableName);
+    Preconditions.checkArgument(!sourceTableLocation().equals(destTableLocation),
+        "Cannot create snapshot where destination location is the same as the source location. This" +
+            "would cause a mixing of original table created and snapshot created files.");
+    this.destTableLocation = destTableLocation;
+  }
+
+  @Override
+  public Long execute() {
+    StagedTable stagedTable;
+    Table icebergTable;
+
+    try {
+      stagedTable = destCatalog().stageCreate(destTableName(), v1SourceTable().schema(),
+          sourcePartitionSpec(), buildPropertyMap());
+      icebergTable = ((SparkTable) stagedTable).table();
+
+      if (!icebergTable.properties().containsKey(TableProperties.DEFAULT_NAME_MAPPING)) {
+        applyDefaultTableNameMapping(icebergTable);
+      }
+    } catch (TableAlreadyExistsException taeException) {
+      throw new IllegalArgumentException("Cannot create snapshot because a table already exists with that name",
+          taeException);
+    } catch (NoSuchNamespaceException nsnException) {
+      throw new IllegalArgumentException("Cannot create snapshot because the namespace given does not exist",
+          nsnException);
+    }
+
+    try {
+      String stagingLocation = icebergTable.location() + "/" + ICEBERG_METADATA_FOLDER;
+      LOG.info("Beginning snapshot of {} to {} using metadata location {}", sourceTableName(), destTableName(),
+          stagingLocation);
+
+      TableIdentifier v1TableIdentifier = v1SourceTable().identifier();
+      SparkTableUtil.importSparkTable(spark(), v1TableIdentifier, icebergTable, stagingLocation);
+      stagedTable.commitStagedChanges();
+
+    } catch (Exception e) {
+      LOG.error("Error when attempting to commit snapshot changes, rolling back", e);
+      if (stagedTable != null) {
+        stagedTable.abortStagedChanges();
+      }
+      throw e;
+    }
+
+    long numMigratedFiles;
+    Snapshot snapshot = icebergTable.currentSnapshot();
+    numMigratedFiles = Long.valueOf(snapshot.summary().get(SnapshotSummary.TOTAL_DATA_FILES_PROP));
+    LOG.info("Successfully loaded Iceberg metadata for {} files", numMigratedFiles);
+    return numMigratedFiles;
+  }
+
+  private Map<String, String> buildPropertyMap() {
+    ImmutableMap.Builder<String, String> propBuilder = ImmutableMap.<String, String>builder()
```

Comment:
Here as well, this will fail if the properties conflict. Since we know that Iceberg won't modify them it is safe to use a mutable map.

Reply:
Sounds good to me